feat: implement improved highlights form with ai auto-summarise

OgDev-01 opened the pull request:
## Description
This PR:
- Adds a new look to the highlights input form for desktop and mobile
- implements auto summarize for highlights 
- adds a new component to the design system
- updates the text area component to include typewriting mode
- improves the overall user experience for creating highlights on mobile and desktop devices

### Notes on summary generation
After testing the various `tone` available, I choose `informative` as it seems to generate a more in-depth context of what was done in the PR in fewer words. `formal` on the other hand was also nice, but it kind of generates almost same words as what was in the commit messages in the PR. I'm open to hearing thoughts on this  🍕 . 

 The generate summary icon is disabled by default and when a request is being made to get the summary from the API to prevent users from clicking when a request is in flight.


## Available tones
` "exciting" | "persuasive" | "informative" | "humorous" | "formal"`


<!-- 
Please do not leave this blank 
This PR [adds/removes/fixes/replaces] the [feature/bug/etc]. 
-->

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
related #1289 
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->

## Mobile & Desktop Screenshots/Recordings

![image](https://github.com/open-sauced/insights/assets/62995161/0508559e-c7ff-4801-b59f-13f626d2ec40)


https://github.com/open-sauced/insights/assets/62995161/f31ba88c-0728-4489-a98e-5bf37f07e313

![image](https://github.com/open-sauced/insights/assets/62995161/79653c76-8b7a-4e10-a556-09e2f1d9de2a)

![image](https://github.com/open-sauced/insights/assets/62995161/3b78de93-4933-42cd-a669-e64c38e312f5)

<!-- Visual changes require screenshots -->


## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?



## [optional] What gif best describes this PR or how it makes you feel?

<img src="https://media4.giphy.com/media/BcPbK9ci4EU31qUTkR/giphy.gif"/>

<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/.github/blob/main/CONTRIBUTING.md.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

brandonroberts commented:
@OgDev-01 nice! Can we provide some feedback that the API request is in flight? Do we prevent them from clicking the suggestion button again while its happening?

OgDev-01 commented:
> @OgDev-01 nice! Can we provide some feedback that the API request is in flight? Do we prevent them from clicking the suggestion button again while its happening?

Opps!, totally missed that. updating shortly... Thanks!

OgDev-01 commented:
Update: I changed the `tone` back to `formal` and reduced `descriptionLength` to `400` as testing with my PR, the description generated were in markdown bullet lists and the length went a above the limit.

bdougie commented:
Some feedback:

1. The generated response consistently comes back with summaries with more than 500 characters. Any chance we can keep it within the limit?
2. The blocking submission UI while waiting is ok, but my suggestion is a making the sparkle icon glow. 
3. I suggest we also generate a title. We can reuse the from the PR or create a prompt to provide a concise title based on the PR generated. This might need an api update to the endpoint to allow a system prompt in addition to the content prompt (git diff and commit messages) - I pitched that here: https://github.com/open-sauced/api/issues/203

## example prompt from - https://github.com/open-sauced/ai/pull/220
![Screen Shot 2023-07-18 at 10 36 57 AM](https://github.com/open-sauced/insights/assets/5713670/1e9c63a9-4380-43d5-93ca-f6c993511f67)

## suggestion
1 & 2 should be addressed in this PR. 3 can just be using the PR title for now. We can do a follow up PR for title generation, unless you think it is approachable.

bdougie commented:
@open-sauced/triage could use some extra eyes on this.

bdougie commented:
Noticed the mobile pencil. Have you tested this with the sitegpt bubble and how they interact?

## mobile highlight pencil
<img width="124" alt="Screen Shot 2023-07-19 at 8 59 11 AM" src="https://github.com/open-sauced/insights/assets/5713670/ca01d206-10f5-4e13-bd20-880149585844">

## sitegpt
<img width="537" alt="Screen Shot 2023-07-19 at 9 00 20 AM" src="https://github.com/open-sauced/insights/assets/5713670/24176dc2-e1a9-4e1f-a07d-7bc9dcff4326">

CBID2 reviewed (commented):
Looks great @OgDev-01.

OgDev-01 commented:
> my suggestion is a making the sparkle icon glow.

Updates: I added a pulse animation to sparkle icon when a request is in flight. Tested with different PRs and none generates more than 500 words for me. @bdougie can you test with the updated changes again?

I changed the `descriptionLength` from `500` to `400` to fix 
> The generated response consistently comes back with summaries with more than 500 characters. Any chance we can keep it within the limit?

OgDev-01 commented:
here is a demo with the new animation effect


https://github.com/open-sauced/insights/assets/62995161/beefb251-5d29-424d-b302-f23a425c5a70

takanome-dev commented:
> > Have you tested this with the sitegpt bubble and how they interact?
> 
> Not sure why, but for some reason, the sitegpt bubble doesn't seem to show up in dev and preview deploys on my own end.

@OgDev-01 yeah it's only visible on prod but you can turn it off to test this out https://github.com/open-sauced/insights/blob/ce126029e01247447be97912cebb6c1cbecca054/pages/_app.tsx#L63-L67

brandonroberts commented on components/atoms/Textarea/text-area.tsx:
```diff
@@ -1,38 +1,64 @@
 interface TextareaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
   defaultRow?: number;
   onChangeText?: (value: string) => void;
+  typewrite?: boolean;
```
I think this should be a separate `typewriter-textarea` atom or `molecule` and the `text-area` atom should remain as is. We only need this typewriter functionality in one place at the moment

brandonroberts commented on lib/utils/generate-pr-highlight-summary.ts:
```diff
@@ -0,0 +1,34 @@
+    if (res.ok) {
+      const data = await res.json();
+      return data.description as string;
+    }
```
Needs a `null` return if no valid response also

brandonroberts commented on lib/utils/github.ts:
```diff
@@ -55,4 +61,32 @@ const generateGhOgImage = (githubUrl: string): { isValid: boolean; url: string }
+const getPullRequestCommitMessageFromUrl = async (url: string): Promise<string[]> => {
+  const [, , , owner, repoName, , pullRequestNumber] = url.split("/");
+
+  const apiUrl = `https://api.github.com/repos/${owner}/${repoName}/pulls/${pullRequestNumber}/commits`;
```
This is going to get rate limited and potentially stop working without a token

bdougie commented:
> Updates: I added a pulse animation to sparkle icon when a request is in flight. Tested with different PRs and none generates more than 500 words for me. bdougie can you test with the updated changes again?

Pulse looks great. 

re: 500 words, I was testing this PR https://github.com/open-sauced/ai/pull/220. But it seems fine now with the "formal" tone change. 

## system prompt

related to this issue on system prompts - https://github.com/open-sauced/api/issues/203

A system prompt is "return any GitHub issues numbers as markdown links." That addition to the prompt would make the generated summary include links instead of plain text issues numbers.

https://deploy-preview-1381--oss-insights.netlify.app/feed/154

<img width="624" alt="Screen Shot 2023-07-20 at 1 13 49 PM" src="https://github.com/open-sauced/insights/assets/5713670/b96cd482-aec7-43fa-8f07-5fe6ff3559ab">

Note the link would need the markdown rendering from https://github.com/open-sauced/insights/pull/1376

OgDev-01 commented:
FYI, with the newly merged change from https://github.com/open-sauced/insights/pull/1396, running `npm run format` had a huge impact on the files changed

OgDev-01 commented:
![image](https://github.com/open-sauced/insights/assets/62995161/31ce86ba-8a59-4acc-a110-f420d16ecc51)
Looking at the overlap in the above screenshot, i'm suggesting hiding the sitegpt action icon in the feeds page.

@bdougie what are your thoughts?

OgDev-01 commented:
Here is another suggestion
![image](https://github.com/open-sauced/insights/assets/62995161/7f7876c8-bcee-4dec-8732-84225ba759f4)

bdougie commented:
> bdougie what are your thoughts?

Let's not go with both. I think that is the best option is hide sitegpt on the highlights page for mobile. I think in the future we should pin the input form to the top of the page on mobile. 

But to be clear just hide the SiteGPT bubble for now. 

![Group 10](https://github.com/open-sauced/insights/assets/5713670/7a14d375-ce29-4194-aad0-4bf7e2b5f2ae)

brandonroberts commented:
@OgDev-01 why so many files changed?

OgDev-01 commented:
> @OgDev-01 why so many files changed?

> FYI, with the newly merged change from https://github.com/open-sauced/insights/pull/1396, running npm run format had a huge impact on the files changed

brandonroberts commented:
Those changed files shouldn't be in your PR if you've merged in from beta though

OgDev-01 commented:
> Those changed files shouldn't be in your PR if you've merged in from beta though

I pulled from beta and got errors while i was trying to push, and then i ran `npm run format` and all those many files where formatted

brandonroberts reviewed (approved):
Functionality works. Extra files changed are from formatting

jpmcb reviewed (approved):
Looking good!!!! 👏🏼 

![Screenshot 2023-07-21 at 3 39 51 PM](https://github.com/open-sauced/insights/assets/23109390/4e5dab18-3948-41e9-b46e-e1b585eb4f4e)
